src/dlfcn.c

```diff
@@ -467,7 +472,157 @@ char *dlerror( void )
+/*
+ * Return adress from Image Allocation Table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static unsigned char *getAddressFromIAT( unsigned char *addr )
+{
+    /* ...inline app code...
+     * 00401002  |. E8 7B0D0000    CALL 00401D82               ; \GetModuleHandleA
+     * ...thunk table...
+     * 00401D82   $-FF25 4C204000 , JMP DWORD PTR DS:[40204C]  ;  KERNEL32.GetModuleHandleA
+     * ...memory address value of pointer...
+     * 40204C > FC 3D 57 7C   ;little endian pointer value
+     */
+    if( addr[0] != 0xff || addr[1] != 0x25 )
+        return NULL;
+
+    HMODULE module = GetModuleHandle( 0 );
+    DWORD size;
+    void **iat = (void **)getImportTable( module, &size );
+    if (!iat)
+        return NULL;
+    /* get offset from thunk table (after instruction 0xff 0x25)
+     *   4018c8 <_VirtualQuery>: ff 25 4a 8a 00 00
+     */
+    ULONG offset = *(ULONG*)(addr+2);
+#ifdef _WIN64
+    /* On 64 bit the offset is relative
+     *   4018c8:	ff 25 4a 8a 00 00    	jmpq   *0x8a4a(%rip)        # 40a318 <__imp_VirtualQuery> # (64bit)
+     */
+    void **ptr = (void *)(addr + 6 + offset);
+#else
+    /* On 32 bit the offset is absolute
+     *    4019b4:	ff 25 90 71 40 00    	jmp    *0x40719
+     */
+    void **ptr = (void *)offset;
+#endif
+    if( ptr < iat || ptr > iat + size )
+        return NULL;
+    return *ptr;
+}
+
+/* holds module filename */
+static char _module_filename[2*MAX_PATH];
+
+/**
+ * Get module information (filename and base address) from the address given
+ * @param addr address to get module info for
+ * @param info pointer to store module info
+ * @return TRUE requested info filled into structure pointed by parameter info
+ * @return FALSE error
+ */
+static BOOL getModuleInfo( const void *addr, Dl_info *info )
+{
+    HMODULE hModule;
+    DWORD sLen;
+
+    if (!GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, addr, &hModule))
+        return FALSE;
+
+    if( !hModule )
+    {
+        return FALSE;
+    }
+
+    info->dli_fbase = (void *)hModule;
+
+    sLen = GetModuleFileNameA( hModule, _module_filename, sizeof( _module_filename ) );
+    if( sLen == 0 )
+        return FALSE;
+    if( sLen == sizeof( _module_filename ) && GetLastError() == ERROR_INSUFFICIENT_BUFFER )
+        return FALSE;
+    info->dli_fname = _module_filename;
+    return TRUE;
+}
+
+DLFCN_EXPORT
+int dladdr( void *addr, Dl_info *info )
+{
+    void *iat_addr;
+    void *real_addr;
+
+    if( !info )
+        return 0;
+
+    iat_addr = getAddressFromIAT( addr );
+    real_addr = iat_addr ? iat_addr : addr;
+    if( !getModuleInfo( real_addr, info ))
+    {
+        info->dli_fname = NULL;
+        info->dli_fbase = NULL;
+        info->dli_saddr = NULL;
+    }
+    else
+    {
+        info->dli_saddr = (void*)real_addr;
+        HMODULE module = GetModuleHandle( 0 );
+        DWORD size;
+        void* iat = getImportTable( module, &size );

```

Comment:
You are calling `getImportTable` two times. Once from `getAddressFromIAT` and second time here. This could be avoided and simplified (by passing iat as argument to `getAddressFromIAT`). Also first argument for `getImportTable` is always `GetModuleHandle(NULL)`. So you do not need to pass this argument and call `GetModuleHandle` directly in `getImportTable`.